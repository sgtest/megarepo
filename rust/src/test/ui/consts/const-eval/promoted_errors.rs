// revisions: noopt opt opt_with_overflow_checks
//[noopt]compile-flags: -C opt-level=0
//[opt]compile-flags: -O
//[opt_with_overflow_checks]compile-flags: -C overflow-checks=on -O

// build-pass
// ignore-pass (test emits codegen-time warnings and verifies that they are not errors)

#![warn(const_err, arithmetic_overflow, unconditional_panic)]

// The only way to have promoteds that fail is in `const fn` called from `const`/`static`.
const fn overflow() -> u32 {
    0 - 1
    //[opt_with_overflow_checks,noopt]~^ WARN any use of this value will cause an error
    //[opt_with_overflow_checks,noopt]~| WARN this was previously accepted by the compiler
}
const fn div_by_zero1() -> i32 {
    1 / 0
    //[opt]~^ WARN any use of this value will cause an error
    //[opt]~| WARN this was previously accepted by the compiler but is being phased out
}
const fn div_by_zero2() -> i32 {
    1 / (1 - 1)
}
const fn div_by_zero3() -> i32 {
    1 / (false as i32)
}
const fn oob() -> i32 {
    [1, 2, 3][4]
}

const X: () = {
    let _x: &'static u32 = &overflow();
    //[opt_with_overflow_checks,noopt]~^ WARN any use of this value will cause an error
    //[opt_with_overflow_checks,noopt]~| WARN this was previously accepted by the compiler
    let _x: &'static i32 = &div_by_zero1();
    //[opt]~^ WARN any use of this value will cause an error
    //[opt]~| WARN this was previously accepted by the compiler but is being phased out
    let _x: &'static i32 = &div_by_zero2();
    let _x: &'static i32 = &div_by_zero3();
    let _x: &'static i32 = &oob();
};

fn main() {}
