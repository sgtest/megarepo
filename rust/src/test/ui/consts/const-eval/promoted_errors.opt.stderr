warning: any use of this value will cause an error
  --> $DIR/promoted_errors.rs:18:5
   |
LL |       1 / 0
   |       ^^^^^
   |       |
   |       attempt to divide `1_i32` by zero
   |       inside `div_by_zero1` at $DIR/promoted_errors.rs:18:5
   |       inside `X` at $DIR/promoted_errors.rs:36:29
...
LL | / const X: () = {
LL | |     let _x: &'static u32 = &overflow();
LL | |
LL | |
...  |
LL | |     let _x: &'static i32 = &oob();
LL | | };
   | |__-
   |
note: the lint level is defined here
  --> $DIR/promoted_errors.rs:9:9
   |
LL | #![warn(const_err, arithmetic_overflow, unconditional_panic)]
   |         ^^^^^^^^^
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>

warning: any use of this value will cause an error
  --> $DIR/promoted_errors.rs:36:28
   |
LL | / const X: () = {
LL | |     let _x: &'static u32 = &overflow();
LL | |
LL | |
LL | |     let _x: &'static i32 = &div_by_zero1();
   | |                            ^^^^^^^^^^^^^^^ referenced constant has errors
...  |
LL | |     let _x: &'static i32 = &oob();
LL | | };
   | |__-
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>

warning: 2 warnings emitted

